### PR TITLE
Drop use of `pbr`

### PR DESCRIPTION
**Pull request by portante**
The `pbr` module is not very helpful to us, since we don't leverage the features it provides [1].

While removing `pbr`, we also address a few problems now exposed, and take advantage of a few opportunities:

 * Add the use of `setuptools_scm` to auto-generate the full list of source files from `git`
 * Correct directories removed in `MANIFEST.in`
   * `pbr` was generating a new version of that file over the one already in our tree
 * Remove `pytest.ini` in favor of the `tool.pytest.ini_options` section in `pyproject.toml`
   * We also drop the pytest config from `setup.cfg`
   * We also move to using `--import-mode=importlib` [2]
 * Use a single `VERSION` file to feed the `setup.cfg` `version` field
 * We correct the `entry_points` option in `setup.cfg` to work with `setuptools`
 * Correct the `tox` `testenv` description

[1] https://docs.openstack.org/pbr/latest/user/features.html
[2] https://docs.pytest.org/en/stable/explanation/pythonpath.html#import-modes

### CODE REVIEW

**Review by dbutenhof — COMMENTED** [review]
Wait: so server and agent are both 0.72 now??

Aside from that, and the fact that you seem to have a broken build dependency here that's not referenced explicitly, dropping `pbr` seems to be surprisingly low-touch.

**Comment by portante** [comment]
> Wait: so server and agent are both 0.72 now??

From my personal view, they have always been that way.  They are in the same repository, and we have one git history for that repository, so a version is just a marker for all code in the repository at that point.

From a `setuptools` point of view, they are as well.  There is only one `pbench` module.  If we had different modules, like `pbench_server` and `pbench_agent, it'd be different.

If you know of a way to package two separate Python3 modules from one repository, I would not mind doing that, despite the fact that I personally see them as one version.  But all that I have read suggests that is not possible.

> Aside from that, and the fact that you seem to have a broken build dependency here that's not referenced explicitly, dropping `pbr` seems to be surprisingly low-touch.

Fixed.

**Comment by dbutenhof** [comment]
> > Wait: so server and agent are both 0.72 now??
> 
> From my personal view, they have always been that way. They are in the same repository, and we have one git history for that repository, so a version is just a marker for all code in the repository at that point.

I find this interesting because my impression had been that you were the original proponent for versioning the agent and server separately so that we ended up planning "Agent 0.71" and "Server 1.0" instead of just "Pbench 0.71"; whereas I'd always thought we should have a single version, even if we're never "releasing" Server 0.71.

And, in the long run, versioning may become mostly meaningless for the server anyway as we move towards continuous deployment.

**Comment by portante** [comment]
> I find this interesting because my impression had been that you were the original proponent for versioning the agent and server separately so that we ended up planning "Agent 0.71" and "Server 1.0" instead of just "Pbench 0.71"; whereas I'd always thought we should have a single version, even if we're never "releasing" Server 0.71.

It depends on the context.  In this GitHub repository, I have been maintaining the v0.72 milestone and v0.72 project for all PRs and issues.

The "goal" of having a Server 1.0 and Agent 1.0 are abstractions maintained elsewhere.

The https://github.com/distributed-system-analysis/pbench/releases/tag/v0.71.0 release is a marked "agent-only" but the tag includes all the code.  We can't get away from that, it is how GitHub makes a release.

If we truly want separate releases for Agent and Server then we need separate repositories.

I'd be in favor of that given that we have very little shared code between them.

> And, in the long run, versioning may become mostly meaningless for the server anyway as we move towards continuous deployment.

Perhaps you mean the semantic versioning thing of `v<major>.<minor>...`, but the Git `SHA1` hashes are an identifier for the version of the code.

**Review by webbnh — DISMISSED** [review]
I have no objections to the changes being made here per se, but, even though this PR is nice and small, it seems like it should (have) be(en) split into three:
- the relocation of the `VERSION` file
- the testing updates (the addition of `lib/pbench/test/functional/server/__init__.py`, the tweak to `tox.ini`, and the migration from `pytest.ini` and `setup.cfg` to the `tool.pytest.ini_options` section in `pyproject.toml`)
- the actual changes to, as the PR title says, drop the use of `pbr`.
E.g, if someone is perusing the PR titles, it's not going to be obvious that _this_ is the PR which reworked the Tox option setup.

Beyond that, I have a nit, some paranoia, and a question:  do we really want to move to a common `VERSION` file?

**Comment by webbnh** [comment]
> The https://github.com/distributed-system-analysis/pbench/releases/tag/v0.71.0 release is a marked "agent-only" but the tag includes all the code. We can't get away from that, it is how GitHub makes a release.

But, GitHub releases don't have much to do with the contents of the `VERSION` files.  I mean, yeah, for an "agent-only" release, we would like the contents of the Agent's `VERSION` file to match the Git(Hub) tag, but other than that, there's not much connection.  (And, as Dave said, we probably won't be doing Server _releases_, especially via GitHub.) 
 However, when we build an RPM (particularly when we put it in one of the new, versioned COPR repos), it's all driven by the `VERSION` file and not by any Git tag...which means that, once this PR goes in, it'll become pretty inconvenient to build a Pbench Server v1.0 RPM (because, it'll come out labeled as v0.72).

Maybe that's OK.  Maybe not.  I'd be interested to hear others' thoughts.

**Comment by portante** [comment]
> But, GitHub releases don't have much to do with the contents of the `VERSION` files.

Inconceivable!  (I do not think it means what you mean.)

> I mean, yeah, for an "agent-only" release, we would like the contents of the Agent's `VERSION` file to match the Git(Hub) tag, but other than that, there's not much connection. 

That just means we don't have the right tooling in place to make this happen.  The tag should be derived from the VERSION file.

> (And, as Dave said, we probably won't be doing Server _releases_, especially via GitHub.)

See my response above.

> However, when we build an RPM (particularly when we put it in one of the new, versioned COPR repos), it's all driven by the `VERSION` file and not by any Git tag...which means that, once this PR goes in, it'll become pretty inconvenient to build a Pbench Server v1.0 RPM (because, it'll come out labeled as v0.72).

As above, that is just mechanism.  The git tag can be created from the `VERSION` file when a release is ready, and that tag used by the "release" mechanism on GitHub.

The fact remains:

Unless we want to write our own packaging and release system to accommodate 2 packages in one repository, there are no tools out there in the Python world to support that.

If the team really wants to separates packages they have to be separate repos with shared code in a third.

This packaging system is too fragile the way it is and in large part because we are trying to maintain the illusion that the existing packaging tools are working with separate packages.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,3 +1,4 @@
+include VERSION
 exclude .coveragerc .flake8 .gitmodules .pre-commit-config.yaml
 prune .github
```
[the remark this answers is not in the document]
Comment (portante) [reply MANIFEST.in:3]: Not sure why, but for whatever reason it is not there in all environments required, unless I add that `include`.

File: server/VERSION [hunk 2]
```diff
@@ -1 +0,0 @@
-1.0.0
```
[the remark this answers is not in the document]
Comment (portante) [reply server/VERSION:1]: Discussion held elsewhere.

**Comment by webbnh** [comment]
> See my response above.

Yeah...it's starting to sound like we need to stop and define what we mean by "version", "package", "packaging", and one or two other terms.  😝

> > However, when we build an RPM (particularly when we put it in one of the new, versioned COPR repos), it's all driven by the `VERSION` file and not by any Git tag...which means that, once this PR goes in, it'll become pretty inconvenient to build a Pbench Server v1.0 RPM (because, it'll come out labeled as v0.72).
> 
> As above, that is just mechanism. The git tag can be created from the `VERSION` file when a release is ready, and that tag used by the "release" mechanism on GitHub.

But, I'm not really interested in the release mechanism on GitHub, and, regardless, it seems to work fine without the `VERSION` file in the top-level directory.  I'm more concerned with how we build RPMs for the Agent and the Server and how they end up with acceptable versions.

> Unless we want to write our own packaging and release system to accommodate 2 packages in one repository, there are no tools out there in the Python world to support that.

Um...I think we are currently managing to do two separate "packages" (whatever those are) from one Git repository today...aren't we?  What tools are we currently (mis)using for that, and how are they not sufficient?

> If the team really wants to separates packages they have to be separate repos with shared code in a third.

I'm pretty sure that that's not true; and, personally, I feel like that's undesirable (not only to me but to you, as well, based on things you've said in the past).

File: server/VERSION (hunk 2 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply server/VERSION:1]: > Discussion held elsewhere.

That discussion has not (yet) touched on _why_ this change was made or what any alternatives might be available.

**Comment by portante** [comment]
> But, I'm not really interested in the release mechanism on GitHub, and, regardless, it seems to work fine without the `VERSION` file in the top-level directory. I'm more concerned with how we build RPMs for the Agent and the Server and how they end up with acceptable versions.

Seems ...

> Um...I think we are currently managing to do two separate "packages" (whatever those are) from one Git repository today...aren't we? What tools are we currently (mis)using for that, and how are they not sufficient?

We are not managing successfully.  We are playing fast and loose with how we construct our packaging that is causing the fragility we are seeing with this change to drop `pbr`.

> I'm pretty sure that that's not true; and, personally, I feel like that's undesirable (not only to me but to you, as well, based on things you've said in the past).

I'd be okay with it if we really want separate agent and server python module packages.  If somebody can solve this problem without introducing lots of bubble gum and baling wire, great.  That person should propose a PR.

**Comment by dbutenhof** [comment]
> > If the team really wants to separates packages they have to be separate repos with shared code in a third.
> 
> I'm pretty sure that that's not true; and, personally, I feel like that's undesirable (not only to me but to you, as well, based on things you've said in the past).

Yeah, that'd be awful. But we ought to be able to build packages based on subdirectory structure. We essentially do that already, albeit with shared infrastructure. If that's not supportable with separate versioning, wow, that's surprising. And although I initially thought it was awkward that `main` was both Agent 0.7x and Server 1.0, I've gotten quite used to that idea and giving it up will seem equally strange.

**Comment by portante** [comment]
No longer working on pbench.

**Pull request closed without merging**